spatialos/UnrealGDK
# Select actor for spatial debugging

#### Description
Allow the client to select an actor for spatial debugging.

- Added a checkbox to the Spatial Debugger pop-up menu to enable spatial debugging for an actor. 
- When this option is enabled, change the mouse to a crosshair and use this position and as the point for sending a raycast into the world. Use the current mouse controls to move the camera and thereby select the desired actor. 
- When a mouse button is pressed with an actor in the crosshair it locks onto the actor and continues to display debug information for that actor, even when it moves from under the crosshair. Therefore, enabling you to select multiple actors at the same time. 
- Clicking again on a selected actor will unlock it and stop displaying debug information for it.
- Cycle through actors under the crosshair. Use the mouse scroll wheel that allows the user to move to the next actor under the crosshair. 
- Visually differentiate what's under the cross-hair by changing the material to a glowing mesh.

[Note : low video quality as 10MB limit here]
![HandoverSelect2Final1hr](https://user-images.githubusercontent.com/64085832/99083524-52f1c880-25ce-11eb-9bcb-515b5d7c5b68.gif)
![TeleportSelectFinal2a](https://user-images.githubusercontent.com/64085832/99082652-1a9dba80-25cd-11eb-8ffd-e6553d9a9987.gif)

Asset changes [NEW]
WBP_SpatialDebuggerConfigMenu (added button "Start Select Actor(s))
![image](https://user-images.githubusercontent.com/64085832/100201625-f1542700-2f08-11eb-92de-1ef9f87608f4.png)
![image](https://user-images.githubusercontent.com/64085832/100201795-252f4c80-2f09-11eb-9057-845ee0138563.png)
![image](https://user-images.githubusercontent.com/64085832/100201880-4132ee00-2f09-11eb-923f-99962892e7a1.png)


Crosshair (New texture)
![image](https://user-images.githubusercontent.com/64085832/97884446-c07c3a00-1d2e-11eb-8749-c371c4b259e9.png)

SpatialDebugger [NEW] (expose the actor types to query, expose the key config, expose highlight option and set default Crosshair texture)
![actortypes-highlighted2](https://user-images.githubusercontent.com/64085832/100202176-a25ac180-2f09-11eb-8d20-be382c819bfe.png)
![NewKeyConfig](https://user-images.githubusercontent.com/64085832/100202193-a7b80c00-2f09-11eb-80b6-dc2d2e2b97b5.png)
![crosshair](https://user-images.githubusercontent.com/64085832/100202407-f06fc500-2f09-11eb-8169-8ca292ccf3cd.png)


GlowingWireframeMaterial (New material)
![image](https://user-images.githubusercontent.com/64085832/99076906-8fb8c200-25c4-11eb-8e91-849ba7302445.png)

#### Release note
- You can now select an actor for spatial debugging in-game. Use F9 (by default) to open the Spatial Debugger in-game config menu and then press the `Start Select Actor(s)` button. Hover over an actor with the mouse to highlight and right-click  (by default) to select. You can select multiple actors. To deselect an actor right-click on it a second time. If there are multiple actors under the cursor use the mouse wheel (by default) to highlight the desired actor then right-click to confirm your selection.

#### Tests
Tested manually in the Handover gym and Teleporting Gym.
Test steps:
- Open a gym in the Unreal Editor.
- Press play.
- Press F9 to open the Spatial Debugger Config Menu.
- Click the "Start Select Actor(s)" button and a crosshair will appear instead of the mouse cursor.
- Hover over an actor and it will be highlighted (glowing mesh)
- Right-click the mouse to select it and you will see debug information displayed above it.
- Right-click the mouse again to de-select it.
- Multiple actors can be selected / deselected.
- If an actor is occluded use the mouse scroll button to highlight the next actors and right-click to select the desired one.

STRONGLY SUGGESTED: How can this be verified by QA?
Follow the test steps documented above.

#### Documentation
Updated the document https://docs.google.com/document/d/1G9cRz891Ycaiax-M7Psx3rlfX8qTIi0CTk8_71PL4Zg/edit#heading=h.j4ppr52lw6ck

Related PR: https://github.com/spatialos/UnrealGDKTestGyms/pull/229
